### ISSUE TITLE
Unadjusted HMC, and generalizing the MH code

### ISSUE THREAD

**Issue by reubenharry** [blackjax-devs/blackjax#747]
### Current behavior

The current HMC implementation performs n steps of discretized Hamiltonian dynamics and then an MH accept/reject.

### Desired behavior

 It would be useful to also have unadjusted HMC (so, the same but without the MH adjustment). This just amounts to running the integrator, so is easy to implement - the question is just whether to have a wrapper around it and present it as a separate algorithm (along with a tuning algorithm in the vein of MCLMC).

Relatedly, if we did that, we could then write standard (i.e. adjusted) HMC as a function of the unadjusted kernel, i.e. in pseudocode:

```python
adjusted_kernel(base_kernel):
    def f(state):
        new_state, info = run_n_steps_of(base_kernel)
        accept = some_function_of(info)
        if accept:
            return new_state
         else: return state 
```

Note that blackjax does something pretty similar already (see `hmc_proposal`).

**Comment by junpenglao** [comment]
Maybe the easiest is to refactor `hmc_proposal`
https://github.com/blackjax-devs/blackjax/blob/b107f9fd60cfc1261a5ce35690b1d0f141041c07/blackjax/mcmc/hmc.py#L255

So user can control the behavior of `sample_proposal`:
1. comparing initial state and last state (current)
2. take the last state (unadjusted HMC)
3. sample from the whole trajectory (https://github.com/blackjax-devs/blackjax/issues/383)

**Comment by junpenglao** [comment]
Actually probably not 3 because there are more efficient way to implement multinomial HMC without storing the full trajectory.

**Comment by reubenharry** [comment]
Yes, this is more of an API question about whether to expose unadjusted HMC as an inference algorithm. Junpeng's suggestion seems reasonable. Ideally, it would be nice if sample_proposal was refactored to be general enough to work for HMC, adjusted MCHMC (on branch) and MALA, in my opinion, but this isn't critical.

**Comment by reubenharry** [comment]
Something to note is that the code for MCLMC (https://github.com/blackjax-devs/blackjax/blob/main/blackjax/mcmc/mclmc.py) differs only very slightly from what you'd do for unadjusted HMC (the integrator is different), so perhaps this should be generalized appropriately
